File: public/python/098-meeting-rooms-ii.py
# LeetCode Problem #253: Meeting Rooms II
# Difficulty: Medium
# Link: https://leetcode.com/problems/meeting-rooms-ii/

from typing import List
import heapq

# ─────────────────────────────────────────────
# APPROACH 1: Min Heap (End Times) | O(n log n) time | O(n) space
# EXPLAIN: Sort by start; use a min-heap of end times — if the earliest-ending room is free, reuse it; otherwise open a new one.
# WHEN: Optimal and standard; directly models room allocation with O(n log n) total.

def solve_1(intervals: List[List[int]]) -> int:
    if not intervals:
        return 0
    intervals.sort()
    min_heap = [intervals[0][1]]
    for i in range(1, len(intervals)):
        if intervals[i][0] >= min_heap[0]:
            heapq.heapreplace(min_heap, intervals[i][1])
        else:
            heapq.heappush(min_heap, intervals[i][1])
    return len(min_heap)

# ─────────────────────────────────────────────
# APPROACH 2: Chronological / Two Sorted Arrays | O(n log n) time | O(n) space
# EXPLAIN: Sort starts and ends separately; advance two pointers — each unmatched start needs a new room.
# WHEN: Elegant two-pointer sweep; avoids heap overhead while keeping the same complexity.

def solve_2(intervals: List[List[int]]) -> int:
    starts = sorted(x[0] for x in intervals)
    ends   = sorted(x[1] for x in intervals)
    rooms = max_rooms = 0
    j = 0
    for i in range(len(starts)):
        if starts[i] < ends[j]:
            rooms += 1
            max_rooms = max(max_rooms, rooms)
        else:
            rooms -= 1
            j += 1
    return max_rooms

# ─────────────────────────────────────────────
# APPROACH 3: Sweep Line (Events) | O(n log n) time | O(n) space
# EXPLAIN: Create +1 events at each start and -1 events at each end; sort and scan to find the peak concurrent count.
# WHEN: Most generalizable — works for any interval scheduling / resource counting problem.

def solve_3(intervals: List[List[int]]) -> int:
    events = []
    for start, end in intervals:
        events.append((start, 1))
        events.append((end, -1))
    # Sort: when times are equal, end (-1) before start (+1) to free room first
    events.sort(key=lambda x: (x[0], x[1]))
    rooms = max_rooms = 0
    for _, delta in events:
        rooms += delta
        max_rooms = max(max_rooms, rooms)
    return max_rooms

# ─────────────────────────────────────────────
# APPROACH 4: Multiset of End Times (Sorted List) | O(n log n) time | O(n) space
# EXPLAIN: Use a sorted list of active end times; for each new interval, remove the earliest end ≤ start if possible.
# WHEN: Mirrors the C++ multiset approach; useful when you need explicit tracking of which rooms end when.

def solve_4(intervals: List[List[int]]) -> int:
    import bisect
    if not intervals:
        return 0
    intervals.sort()
    end_times = [intervals[0][1]]
    for i in range(1, len(intervals)):
        start, end = intervals[i]
        # Find the earliest room that ends at or before current start
        idx = bisect.bisect_right(end_times, start) - 1
        if idx >= 0:
            end_times.pop(idx)
        bisect.insort(end_times, end)
    return len(end_times)

# ─────────────────────────────────────────────
# APPROACH 5: Priority Queue with (end, start) Pairs | O(n log n) time | O(n) space
# EXPLAIN: Push (end_time, start_time) into a min-heap; pop all rooms whose end ≤ current start before pushing.
# WHEN: Variant of Approach 1 that pops multiple expired rooms; useful when batch-freeing rooms matters.

def solve_5(intervals: List[List[int]]) -> int:
    if not intervals:
        return 0
    intervals.sort()
    heap: list = []
    max_rooms = 0
    for start, end in intervals:
        while heap and heap[0][0] <= start:
            heapq.heappop(heap)
        heapq.heappush(heap, (end, start))
        max_rooms = max(max_rooms, len(heap))
    return max_rooms

# Made with Bob

File: public/java/098-meeting-rooms-ii.java
/*
 * LeetCode Problem #253: Meeting Rooms II
 * Link: https://leetcode.com/problems/meeting-rooms-ii/
 * Difficulty: Medium
 */
import java.util.*;

class Solution {

    // APPROACH 1: Min Heap (End Times) | O(n log n) time | O(n) space
    // EXPLAIN: Sort by start; use a min-heap of end times — reuse the earliest-ending room if free.
    public int minMeetingRooms1(int[][] intervals) {
        if (intervals.length == 0) return 0;
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        minHeap.offer(intervals[0][1]);
        for (int i = 1; i < intervals.length; i++) {
            if (intervals[i][0] >= minHeap.peek()) minHeap.poll();
            minHeap.offer(intervals[i][1]);
        }
        return minHeap.size();
    }

    // APPROACH 2: Chronological / Two Sorted Arrays | O(n log n) time | O(n) space
    // EXPLAIN: Sort starts and ends separately; advance two pointers — each unmatched start needs a new room.
    public int minMeetingRooms2(int[][] intervals) {
        int n = intervals.length;
        int[] starts = new int[n], ends = new int[n];
        for (int i = 0; i < n; i++) { starts[i] = intervals[i][0]; ends[i] = intervals[i][1]; }
        Arrays.sort(starts); Arrays.sort(ends);
        int rooms = 0, maxRooms = 0, j = 0;
        for (int i = 0; i < n; i++) {
            if (starts[i] < ends[j]) { rooms++; maxRooms = Math.max(maxRooms, rooms); }
            else { rooms--; j++; }
        }
        return maxRooms;
    }

    // APPROACH 3: Sweep Line (Events) | O(n log n) time | O(n) space
    // EXPLAIN: Create +1 events at starts and -1 at ends; sort and scan for the peak concurrent count.
    public int minMeetingRooms3(int[][] intervals) {
        int n = intervals.length;
        int[][] events = new int[2 * n][2];
        for (int i = 0; i < n; i++) {
            events[2 * i]     = new int[]{intervals[i][0],  1};
            events[2 * i + 1] = new int[]{intervals[i][1], -1};
        }
        Arrays.sort(events, (a, b) -> a[0] != b[0] ? a[0] - b[0] : a[1] - b[1]);
        int rooms = 0, maxRooms = 0;
        for (int[] e : events) { rooms += e[1]; maxRooms = Math.max(maxRooms, rooms); }
        return maxRooms;
    }

    // APPROACH 4: TreeMap (Sweep Line with Counts) | O(n log n) time | O(n) space
    // EXPLAIN: Use a TreeMap to record +1 at each start and -1 at each end; scan in key order for peak.
    public int minMeetingRooms4(int[][] intervals) {
        TreeMap<Integer, Integer> diff = new TreeMap<>();
        for (int[] iv : intervals) {
            diff.merge(iv[0],  1, Integer::sum);
            diff.merge(iv[1], -1, Integer::sum);
        }
        int rooms = 0, maxRooms = 0;
        for (int delta : diff.values()) { rooms += delta; maxRooms = Math.max(maxRooms, rooms); }
        return maxRooms;
    }

    // APPROACH 5: Priority Queue with (end, start) Pairs | O(n log n) time | O(n) space
    // EXPLAIN: Push (endTime, startTime) into a min-heap; pop expired rooms before counting for each meeting.
    public int minMeetingRooms5(int[][] intervals) {
        if (intervals.length == 0) return 0;
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        int maxRooms = 0;
        for (int[] iv : intervals) {
            while (!pq.isEmpty() && pq.peek()[0] <= iv[0]) pq.poll();
            pq.offer(new int[]{iv[1], iv[0]});
            maxRooms = Math.max(maxRooms, pq.size());
        }
        return maxRooms;
    }
}

// Made with Bob
